Add inference code to MNIST example (#4741)
chainer/chainer: examples/mnist/README.md

   `-- snapshot_iter_12000
```

`cg.dot` is a Graphviz dot file that contains the graph structure of the neural network used in this training script.
`log` is a JSON file that contains information during training.
`snapshot_iter_12000` is a snapshot of `Trainer` object which enables to resume the training.
You can also extract the trained model parameters from this snapshot file.

If you have installed [matplotlib](https://matplotlib.org/) before the training, the following additional files will saved in the `result` dir:

```
accuracy.png
loss.png
```

These images visualize the accuracies and loss values changing through time during training.

### `train_mnist_custom_loop.py`

This example shows how you can write the training loop by yourself without using `Trainer` class abstraction.

Usage (with a GPU):
```
./train_mnist_custom_loop.py -g 0
```

Usage (with CPUs):
```
./train_mnist_custom_loop.py
```

Then, the following files will be created:

```
-- result
   |-- mlp.model
   `-- mlp.state
```

`mlp.model` contains the trained parameters of the model.
`mlp.state` contains the internal states of Adam optimizer after the training.

### `train_mnist_data_parallel.py`

> This example shows how to use multiple GPUs with `Trainer` class in data-parallel manner.

Reviewer: ```suggestion
This example shows how to use multiple GPUs with the `Trainer` class in a data-parallel manner.
```

> This script does not support CPUs and requires at least two GPUs.

Reviewer: ```suggestion
This script does not support CPUs and **requires at least two GPUs.**
```

Usage:
```
./train_mnist_data_parallel.py
```

> Then, the following files will be craeted:

Reviewer: ```suggestion
Then, the following files will be created:
```

```
-- result_data_parallel
   |-- cg.dot
   |-- log
   `-- snapshot_iter_3000
```

### `train_mnist_model_parallel.py`

This example shows how to split a single model to different GPUs and train it with `Trainer` class.
This script does not support CPUs and **requires at least two GPUs.**

Usage:
```
./train_mnist_model_parallel.py
```

Then, the following files will be craeted:

```
-- result_model_parallel
   |-- cg.dot
   |-- log
   `-- snapshot_iter_12000
```

## Inference

> `inference.py` shows how to load the saved parameters which is trained by one of the above training scripts and use it for making a prediction on a test datum. It requires a saved snapshot file containing learnt parameters, so please run one of the above training script first.

Reviewer: ```suggestion
`inference.py` shows how to load the saved parameters trained by one of the above training scripts and use it for making a prediction on a test datum. It requires a saved snapshot file containing learned parameters, so please run one of the above training script first.
```
